### PR TITLE
[internal/pdatautil] Change MapHash result from 64 to 128 bit

### PR DESCRIPTION
**Pull request by dmitryax**
For higher collision tolerance

Benchmark results:

```
goos: darwin
goarch: arm64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/pdatautil
BenchmarkMapHashFourItems
BenchmarkMapHashFourItems-10                  	 3161750	       377.5 ns/op	     272 B/op	      10 allocs/op
BenchmarkMapHashEightItems
BenchmarkMapHashEightItems-10                 	 1552203	       782.6 ns/op	     440 B/op	      18 allocs/op
BenchmarkMapHashWithEmbeddedSliceAndMap
BenchmarkMapHashWithEmbeddedSliceAndMap-10    	  929382	      1228 ns/op	     768 B/op	      31 allocs/op
PASS
```

### CODE REVIEW

File: internal/coreinternal/pdatautil/map.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	valIntPrefix    = []byte{'\xfa'}
+	valDoublePrefix = []byte{'\xfb'}
+	valMapPrefix    = []byte{'\xfc'}
+	valMapSuffix    = []byte{'\xfd'}
+	valSlicePrefix  = []byte{'\xfe'}
+	valSliceSuffix  = []byte{'\xff'}
+)
+
+// MapHash return a hash for the provided map.
+// Order of the underlying key/value pairs is ignored.
```
Comment (atoulme) [inline internal/coreinternal/pdatautil/map.go:28]: if the order is ignored, does this mean the hash is not deterministic? or that we discard the order to use the internal order of elements are stored in the underlying storage.

Comment (atoulme) [reply internal/coreinternal/pdatautil/map.go:28]: actually, we sort by keys line 41, so we do care about the order. Should this comment be removed?

Comment (dmitryax) [reply internal/coreinternal/pdatautil/map.go:28]: The order is ignored specifically to make it deterministic. I can rephrase it to make it clear

Comment (dmitryax) [reply internal/coreinternal/pdatautil/map.go]: Updated

Comment (atoulme) [reply internal/coreinternal/pdatautil/map.go]: :+1:

File: internal/coreinternal/pdatautil/map.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	valDoublePrefix = []byte{'\xfb'}
+	valMapPrefix    = []byte{'\xfc'}
+	valMapSuffix    = []byte{'\xfd'}
+	valSlicePrefix  = []byte{'\xfe'}
+	valSliceSuffix  = []byte{'\xff'}
+)
+
+// MapHash return a hash for the provided map.
+// Maps with the same underlying key/value pairs in different order produce the same deterministic hash value.
+func MapHash(m pcommon.Map) uint64 {
```
Comment (bogdandrutu) [inline internal/coreinternal/pdatautil/map.go:43]: I would go with 16 bytes array (not slice) as return type. Allows to use UUID or other implementation.

File: internal/coreinternal/pdatautil/map.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	h := xxhash.New()
+	writeMapHash(h, m)
+	return h.Sum64()
+}
+
+func writeMapHash(h hash.Hash, m pcommon.Map) {
```
Comment (bogdandrutu) [inline internal/coreinternal/pdatautil/map.go:49]: Would like to see a benchmark of this implementation for a classic case like 4/8/16 arguments

File: internal/coreinternal/pdatautil/map.go (hunk 2 above)
Comment (dmitryax) [reply internal/coreinternal/pdatautil/map.go:43]: What I liked about uint64 is that results can be easily comparable but that maybe doesn't have colossal collision tolerance

File: internal/coreinternal/pdatautil/map.go (hunk 3 above)
Comment (dmitryax) [reply internal/coreinternal/pdatautil/map.go:57]: Added benchmarks:

```
goos: darwin
goarch: arm64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/pdatautil
BenchmarkMapHashFourItems
BenchmarkMapHashFourItems-10                  	 3161750	       377.5 ns/op	     272 B/op	      10 allocs/op
BenchmarkMapHashEightItems
BenchmarkMapHashEightItems-10                 	 1552203	       782.6 ns/op	     440 B/op	      18 allocs/op
BenchmarkMapHashWithEmbeddedSliceAndMap
BenchmarkMapHashWithEmbeddedSliceAndMap-10    	  929382	      1228 ns/op	     768 B/op	      31 allocs/op
PASS
```

I'll also try SHA1 to generate uuidv5 to see the difference

File: internal/coreinternal/pdatautil/map.go (hunk 2 above)
Comment (dmitryax) [reply internal/coreinternal/pdatautil/map.go]: Changed to [16]byte

File: internal/coreinternal/pdatautil/map.go (hunk 3 above)
Comment (dmitryax) [reply internal/coreinternal/pdatautil/map.go:57]: SHA1 implementation just a bit slower:

```golang
func MapHash(m pcommon.Map) [sha1.Size]byte {
	h := sha1.New()
	writeMapHash(h, m)
	b := h.Sum(nil)
	res := [sha1.Size]byte{}
	copy(res[:], b)
	return res
}
```

```
goos: darwin
goarch: arm64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/pdatautil
BenchmarkMapHashFourItems
BenchmarkMapHashFourItems-10                  	 2616472	       452.1 ns/op	     328 B/op	      11 allocs/op
BenchmarkMapHashEightItems
BenchmarkMapHashEightItems-10                 	 1315567	       907.3 ns/op	     496 B/op	      19 allocs/op
BenchmarkMapHashWithEmbeddedSliceAndMap
BenchmarkMapHashWithEmbeddedSliceAndMap-10    	  795008	      1496 ns/op	     824 B/op	      32 allocs/op
PASS
```

Comment (dmitryax) [reply internal/coreinternal/pdatautil/map.go]: Do you think we should switch to sha1/uuid5?

**Comment by dmitryax** [comment]
@bogdandrutu, the first version of this function got merged as part of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17179. This PR is a diff from that state now

**Review by bogdandrutu — APPROVED** [review]

**Review by djaglowski — APPROVED** [review]
I ran the benchmarks against `main` for comparison. Looks good.

```
// main
goos: darwin
goarch: arm64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/pdatautil
BenchmarkMapHashFourItems-10                  	 3477991	       332.6 ns/op	     272 B/op	      10 allocs/op
BenchmarkMapHashEightItems-10                 	 1741002	       685.1 ns/op	     440 B/op	      18 allocs/op
BenchmarkMapHashWithEmbeddedSliceAndMap-10    	 1000000	      1150 ns/op	     768 B/op	      31 allocs/op

// pr
goos: darwin
goarch: arm64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/pdatautil
BenchmarkMapHashFourItems-10                  	 3367083	       344.2 ns/op	     272 B/op	      10 allocs/op
BenchmarkMapHashEightItems-10                 	 1716458	       697.5 ns/op	     440 B/op	      18 allocs/op
BenchmarkMapHashWithEmbeddedSliceAndMap-10    	 1000000	      1167 ns/op	     768 B/op	      31 allocs/op
```
